ignition_ros2_control/src/ignition_system.cpp

```diff
@@ -0,0 +1,402 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "ignition_ros2_control/ignition_system.hpp"
+
+#include "ignition/gazebo/components/JointForce.hh"
+#include "ignition/gazebo/components/JointForceCmd.hh"
+#include "ignition/gazebo/components/JointPosition.hh"
+#include "ignition/gazebo/components/JointPositionReset.hh"
+#include "ignition/gazebo/components/JointVelocity.hh"
+#include "ignition/gazebo/components/JointVelocityCmd.hh"
+
+class ignition_ros2_control::IgnitionSystemPrivate
+{
+public:
+  IgnitionSystemPrivate() = default;
+
+  ~IgnitionSystemPrivate() = default;
+  /// \brief Degrees od freedom.
+  size_t n_dof_;
+
+  /// \brief last time the write method was called.
+  rclcpp::Time last_update_sim_time_ros_;
+
+  /// \brief vector with the joint's names.
+  std::vector<std::string> joint_names_;
```

Comment:
This current implementation relies on various vectors being the same size and storing joint information in the same order. I think this will probably be true most of the time, but any error could be tricky to spot.

I'd suggest creating a struct to hold all the information you need for each joint, and storing a single vector with these structs. I think that would make it easier to manage all the information.